packages/firestore/src/api/cache_config.ts

```diff
@@ -0,0 +1,302 @@
+ * set the instance to `FirestoreSettings.cache` and call `initializeFirestore` using
+ * the settings object.
+ */
+export type MemoryLocalCache = {
+  kind: 'memory';
+  /**
+   * @internal
+   */
+  _onlineComponentProvider: OnlineComponentProvider;
+  /**
+   * @internal
+   */
+  _offlineComponentProvider: MemoryOfflineComponentProvider;
+};
+
+class MemoryLocalCacheImpl implements MemoryLocalCache {
+  kind: 'memory' = 'memory';
+  /**
+   * @internal
+   */
+  _onlineComponentProvider: OnlineComponentProvider;
+  /**
+   * @internal
+   */
+  _offlineComponentProvider: MemoryOfflineComponentProvider;
+
+  constructor() {
+    this._onlineComponentProvider = new OnlineComponentProvider();
+    this._offlineComponentProvider = new MemoryOfflineComponentProvider();
+  }
+
+  toJSON(): {} {
+    return { kind: this.kind };
+  }
+}
+
+/**
+ * Provides a cache backed by IndexedDb to the SDK.
+ *
+ * To use, create an instance using the factory function {@link indexedDbLocalCache()}, then
+ * set the instance to `FirestoreSettings.cache` and call `initializeFirestore` using
+ * the settings object.
+ */
+export type IndexedDbLocalCache = {
+  kind: 'indexeddb';
+  /**
+   * @internal
+   */
+  _onlineComponentProvider: OnlineComponentProvider;
+  /**
+   * @internal
+   */
+  _offlineComponentProvider: OfflineComponentProvider;
+};
+
+class IndexedDbLocalCacheImpl implements IndexedDbLocalCache {
+  kind: 'indexeddb' = 'indexeddb';
+  /**
+   * @internal
+   */
+  _onlineComponentProvider: OnlineComponentProvider;
+  /**
+   * @internal
+   */
+  _offlineComponentProvider: OfflineComponentProvider;
+
+  constructor(settings: IndexedDbCacheSettings | undefined) {
+    let tabManager: IndexedDbTabManager;
+    if (settings?.tabManager) {
+      settings.tabManager._initialize(settings);
+      tabManager = settings.tabManager;
+    } else {
+      tabManager = indexedDbSingleTabManager(undefined);
+      tabManager._initialize(settings);
+    }
+    this._onlineComponentProvider = tabManager._onlineComponentProvider!;
+    this._offlineComponentProvider = tabManager._offlineComponentProvider!;
+  }
+
+  toJSON(): {} {
+    return { kind: this.kind };
+  }
+}
+
+/**
+ * Union type from all supported SDK cache layer.
+ */
+export type FirestoreLocalCache = MemoryLocalCache | IndexedDbLocalCache;
+
+/**
+ * Creates an instance of `MemoryLocalCache`. The instance can be set to
+ * `FirestoreSettings.cache` to tell the SDK what cache layer to use.
+ */
+export function memoryLocalCache(): MemoryLocalCache {
+  return new MemoryLocalCacheImpl();
+}
+
+/**
+ * An settings object to configure an `IndexedDbLocalCache` instance.
+ */
+export type IndexedDbCacheSettings = {
+  /**
+   * An approximate cache size threshold for the on-disk data. If the cache
+   * grows beyond this size, Firestore will start removing data that hasn't been
+   * recently used. The size is not a guarantee that the cache will stay below
```

Comment:
The SDK does not guarantee that the cache...

Reply:
Done.